### DISCUSSION TITLE
How can I keep a "local" tf module next to/close to the terragrunt.hcl?

### DISCUSSION THREAD

**Discussion by pieterbreed** [gruntwork-io/knowledge-base#559] [Help] [answered]
Hi, I'm fairly new to terragrunt, so apologies if I'm asking something with an obvious answer.

I'm confused by this section in the [example tutorial](https://terragrunt.gruntwork.io/docs/getting-started/quick-start/):

```
stage
├── terragrunt.hcl
├── frontend-app
│   ├── main.tf
│   └── terragrunt.hcl
└── mysql
    ├── main.tf
    └── terragrunt.hcl
```

This seems to hint at a capability what I want, but I can't get it to work.

I have some terraform modules that are fairly generic and shared over some terragrunt directories. I have them on the same filesystem (not in git). It looks more or less like this:

```
├── terraform
│   └── modules
│       ├── application-cache
│       │   ├── main.tf
│       ├── centralized-alarms
│       │   ├── main.tf
...
└── terragrunt
    ├── prodapp
    │   ├── account.yaml
    │   ├── appcache
    │   │   ├── local
    │   │   │   ├── main.tf
    │   │   └── terragrunt.hcl
```

For one of terragrunt.hcl files, I need a little bit of local customization in how a shared module (`application-cache`) is called. Specifically, one of the parameters that I just provide manually in the testing environment, instead needs to be read from a cloudformation output. So I have a little bit of local, environment-specific customization around invoking the shared/generic module. 

My problem is I cannot figure out how to achieve this. I don't want to create another module in the "shared" modules directory which is specifically only serves as an "entry point" module for one specific environment. I'd rather keep the terraform code "close" to the terragrunt.hcl it applies to.

I've tried to have the `main.tf` in the same folder as the `terragrunt.hcl` file like the example tutorial, which seems to imply that I must not have a `terraform { source ... }` block in `terragrunt.hcl`. But then the locals are not passed correctly to the variables I declare in the `main.tf` using the `input` block. When I run `terragrunt apply` it asks for the value of a variable that is declared in the `main.tf` and specified in the `terragrunt.hcl` file's `inputs`.

I've also tried having a `./local` directory with `./local/main.tf`, attempting to treat it as a terraform module. In the `terragrunt.hcl` file I then specify `terraform { source = ".//local" }` bit this fails with paths to the shared terraform modules not being resolved correctly.

```
| Error: Unreadable module directory
│ 
│ Unable to evaluate directory symlink: lstat ../../../../terraform: no such
│ file or directory
```

Really my question is, if I have shared terraform modules, and my environments have slight dissimilarities in how those shared modules are invoked, how must I structure the terragrunt repository to accomplish this?

---

<ins datetime="2022-09-19T11:55:36Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109257">Tracked in ticket #109257</a></p>
</ins>

**Comment by denis256** [comment, marked as answer]
Hi,
AFAIK, it depends on how big are dissimilarities:
 * can be used `dependency` block and outputs passed to `application-cache` - but it will require to extract terraform code as separated module and reference it in Terragrunt
 * can be used `generate` block to put custom code in the module directory before the invocation of Terraform

References:
https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#dependency
https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#generate

I also created some examples using `generate` and `dependency` in https://github.com/denis256/terragrunt-tests/tree/master/discussion-559

I hope this helps

**Reply by pieterbreed** [reply]
Thank you for the suggestions. I'll have to read the docs with attention and think carefully if they provide the capability that I need.

**Comment by ivellios** [comment]
I am having a similar issue, thus I won't create a new discussion and put it here:

In my child `env/stage/vpc/terragrunt.hcl` I have:

```hcl
terraform {
  source = "${path_relative_from_include()}/modules//vpc"  # also tried: source = "../../../modules/vpc" with the same effect
}
```

In the root directory I also have `terragrunt.hcl` file, so it is on the same level as `modules` directory. When running terragrunt I see the modules showing up in the `.terragrunt-cache`. But still, it looks as if the module was not being run and inputs passed into it. I get the message that there's no change in the infrastructure.

But if I remove the above code and put `env/stage/vpc/main.tf` in the same child directory with terraform `module` and source param pointing to the local module, it is being used.

Terraform `main.tf` file:
```hcl
module "vpc" {
  source = "../../../modules/vpc"
  ...
```

The simplified files structure:
```
| - env - stage - vpc
|                  | - terragrunt.hcl
|                  | - main.tf
|                  | - ...
| - modules - vpc
|              | - main.tf
| - terragrunt.hcl
```


How can I get rid of this `main.tf` file and use the module directly from the `terragrunt.hcl`? Am I missing something here?

**Disclaimer:** I am new to terragrunt, sorry!

**Comment by ivellios** [comment]
Looks like it wasn't working because of the `privder/backend/versions.tf` files present in the `env-stage-vpc` directory. I removed them and moved the module to another directory and it worked like a charm. Also no files were visibly generated by terragrunt (which is quite unexpected, but also makes it clearer to use)
